packages/react/src/experimental/OneCalendar/OneCalendar.tsx

```diff
@@ -89,20 +91,116 @@ export function OneCalendar({
+    to: boolean
+  }>({
+    from: false,
+    to: false,
+  })
+
+  const handleInputChange = (input: "from" | "to") => {
+    setSelectFromInput(input, inputValue)
+  }
+
+  const setSelectFromInput = (
+    input: "from" | "to",
+    inputValue: DateRangeString
+  ) => {
+    const newDate = granularity.fromString(inputValue)
+    const error = newDate && newDate[input] && !isValidDate(newDate[input])
+    setInputError((prev) => ({
+      ...prev,
+      [input]: error,
+    }))
+
+    if (!error) {
+      setSelected(newDate)
+    }
+  }
+  useEffect(() => {
+    const range = toDateRange(selected)
+
+    const { from, to } = granularity.toRangeString(
+      range ? range : { from: new Date(), to: undefined }
+    )
+    setInputValue({
+      from: from || "",
+      to: to || "",
+    })
+  }, [granularity, selected])
+
+  const handleInputNavigate = (input: "from" | "to", direction: -1 | 1) => {
+    const currentDate = inputValue[input]
+      ? granularity.fromString(inputValue[input])
+      : undefined
+    const newDate = currentDate
+      ? granularity.navigate(currentDate.from, direction)
+      : undefined
+
+    if (isValidDate(newDate)) {
+      const newInputValue = {
+        ...inputValue,
+        [input]: granularity.toRangeString(newDate).from,
+      }
+      setInputValue(newInputValue)
+      setSelectFromInput(input, newInputValue)
+    }
+  }
 
   return (
     <div className="flex flex-col">
       {showInput && (
-        <div className="mb-2">
-          <Input value={dateString} disabled />
+        <div className="mb-2 flex gap-2">
+          <Input
+            error={!!inputError.from}
+            value={inputValue.from}
+            placeholder={mode === "range" ? i18n.date.from : i18n.date.date}
+            onBlur={() => handleInputChange("from")}
+            onKeyDown={(e) => {
+              if (e.key === "Enter") {
+                handleInputChange("from")
+              }
+              if (e.key === "ArrowUp" || e.key === "ArrowDown") {
+                e.preventDefault()
+                handleInputNavigate("from", e.key === "ArrowDown" ? -1 : 1)
+              }
+            }}
+            onChange={(e) =>
+              setInputValue({ ...inputValue, from: e.target.value })
+            }
+          />
+          {mode === "range" && (
+            <Input
+              error={!!inputError.to}
+              value={inputValue.to}
+              placeholder={i18n.date.to}
+              onBlur={() => handleInputChange("to")}
+              onKeyDown={(e) => {
+                if (e.key === "Enter") {
+                  handleInputChange("to")
+                }
+                if (e.key === "ArrowUp" || e.key === "ArrowDown") {
+                  e.preventDefault()
+                  handleInputNavigate("to", e.key === "ArrowDown" ? -1 : 1)
+                }
+              }}
+              onChange={(e) =>
+                setInputValue({ ...inputValue, to: e.target.value })
+              }
```

Comment:
I would move this to a no inline arrow function to improve the clarity of the code